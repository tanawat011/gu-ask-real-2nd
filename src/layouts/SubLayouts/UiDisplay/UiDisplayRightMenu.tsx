import tw, { styled } from 'twin.macro'

import { RightMenuContent } from 'components/PageUiComponent'

export type UiDisplayRightMenuItems = {
  [key: string]: {
    id: string
    name: string
  }
}

export type UiDisplayRightMenuProps = {
  isScrolled: boolean
  containerId: string
  menuItems: UiDisplayRightMenuItems
}

const TwContainer = styled.div(({ scrolled }: { scrolled: boolean }) => {
  return [tw`z-10`, scrolled && tw`fixed`]
})

export const UiDisplayRightMenu: React.FC<UiDisplayRightMenuProps> = (
  props,
) => {
  const _menuList = Object.values(props.menuItems).map((item) => item.name)

  return (
    <TwContainer scrolled={props.isScrolled}>
      <RightMenuContent
        containerId={props.containerId}
        title='TABLE OF CONTENT'
        menuList={_menuList}
      />
    </TwContainer>
  )
}
