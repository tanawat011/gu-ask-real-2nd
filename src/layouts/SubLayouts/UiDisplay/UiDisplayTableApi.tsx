import type { ComponentApiGuideData } from 'types'

import React from 'react'

import tw from 'twin.macro'

import { Title } from 'components/PageUiComponent'
import { TableCommon } from 'components/Table'
import { componentApiGuideColumns } from 'data'

const TwContainer = tw.div`mb-8`

export type UiDisplayTableApiProps = {
  apiId: string
  apiName: string
  apiData: ComponentApiGuideData<unknown>[]
}

export const UiDisplayTableApi: React.FC<UiDisplayTableApiProps> = (props) => {
  return (
    <TwContainer id={props.apiId}>
      <Title title={props.apiName} />

      <TableCommon columns={componentApiGuideColumns} data={props.apiData} />
    </TwContainer>
  )
}
