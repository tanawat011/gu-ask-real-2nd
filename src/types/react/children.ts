export type Children = { children: JSX.Element }
