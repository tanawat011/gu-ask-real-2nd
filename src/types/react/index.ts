export * from './children'
export * from './reactRouterDom'
