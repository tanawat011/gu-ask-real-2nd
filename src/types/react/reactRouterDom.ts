import type { Scroll } from 'types/common'

export type OutletContextProps = {
  scroll: Scroll
}
