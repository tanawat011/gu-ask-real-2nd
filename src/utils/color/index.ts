export * from './hex2Rgb'
export * from './rgb2Hex'
