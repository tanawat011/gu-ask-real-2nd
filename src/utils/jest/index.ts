export * from './renderWithProviders'
export * from './twColor'
