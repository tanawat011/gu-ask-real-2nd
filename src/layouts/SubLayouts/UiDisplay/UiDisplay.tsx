import type { UiDisplayRightMenuProps } from './UiDisplayRightMenu'
import type { UiDisplayTableApiProps } from './UiDisplayTableApi'

import React, { useState } from 'react'

import { Outlet } from 'react-router-dom'
import tw from 'twin.macro'

import { Title } from 'components/PageUiComponent'

import { UiDisplayRightMenu } from './UiDisplayRightMenu'
import { UiDisplayTableApi } from './UiDisplayTableApi'

const TwContainer = tw.div`relative grid grid-cols-1 gap-8 overflow-x-auto px-8 py-6 desktop:grid-cols-5`
const TwWrapContent = tw.div`xl:col-span-4`
const TwWrapMenu = tw.div`col-span-1 hidden desktop:block`

export type UiDisplayLayoutOutletContext = {
  handleChildProps: (props: UiDisplayLayoutChildProps) => void
}

export type UiDisplayLayoutChildProps = {
  title: string
  description: string
} & UiDisplayTableApiProps &
  Omit<UiDisplayRightMenuProps, 'isScrolled'>

const defaultChildProps: UiDisplayLayoutChildProps = {
  title: '',
  description: '',
  containerId: '',
  apiId: '',
  apiName: '',
  apiData: [],
  menuItems: {},
}

export const UiDisplayLayout = () => {
  const [childProps, setChildProps] =
    useState<UiDisplayLayoutChildProps>(defaultChildProps)
  const [isScrolled, setIsScrolled] = useState(false)

  const handleScroll = () => {
    setIsScrolled(true)
  }

  const handleChildProps = (props: UiDisplayLayoutChildProps) => {
    setChildProps(props)
  }

  return (
    <TwContainer
      id={childProps.containerId}
      onScroll={handleScroll}
      data-testid='ui-display-container-layout'
    >
      <TwWrapContent>
        <div>
          <Title
            size='xl'
            title={childProps.title}
            description={childProps.description}
          />
        </div>

        <Outlet
          context={{ handleChildProps } as UiDisplayLayoutOutletContext}
        />

        <UiDisplayTableApi {...childProps} />
      </TwWrapContent>

      <TwWrapMenu>
        <UiDisplayRightMenu {...childProps} isScrolled={isScrolled} />
      </TwWrapMenu>
    </TwContainer>
  )
}
