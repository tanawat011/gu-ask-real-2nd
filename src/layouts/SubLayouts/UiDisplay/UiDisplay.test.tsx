import { fireEvent, screen } from '@testing-library/react'
import { Route, Routes, useOutletContext } from 'react-router-dom'

import { renderWithProviders } from 'utils/jest'

import { UiDisplayLayout } from './UiDisplay'

describe('<UiDisplayLayout />', () => {
  test('should render', () => {
    expect.assertions(1)

    const { container } = renderWithProviders(<UiDisplayLayout />)

    expect(container).toBeInTheDocument()
  })

  test('match snapshot', () => {
    expect.assertions(1)

    window.location.assign('http://localhost:3000/ui-component')
    const { asFragment } = renderWithProviders(<UiDisplayLayout />, {
      router: { route: '/ui-component' },
    })

    expect(asFragment()).toMatchSnapshot()
  })

  test('should can handle scroll event and handle child prop successful', () => {
    expect.assertions(1)

    const FakeComponent = () => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const { handleChildProps } = useOutletContext<any>()

      return (
        <div
          onClick={() =>
            handleChildProps({
              title: '',
              description: '',
              containerId: '',
              apiId: '',
              apiName: '',
              apiData: [],
              menuItems: {
                basic: { id: 'basic', name: 'basic' },
              },
            })
          }
          data-testid='fake-component'
        >
          Fake Component
        </div>
      )
    }

    window.location.assign('http://localhost:3000/ui-component/test')
    const { asFragment } = renderWithProviders(
      <Routes>
        <Route path='/ui-component' element={<UiDisplayLayout />}>
          <Route path='test' element={<FakeComponent />} />
        </Route>
      </Routes>,
      {
        router: { route: '/ui-component/test' },
      },
    )

    const container = screen.getByTestId(/ui-display-container-layout/i)
    const fakeComponent = screen.getByTestId(/fake-component/i)

    fireEvent.scroll(container, { target: { scrollY: 100 } })
    fireEvent.click(fakeComponent)

    expect(asFragment()).toMatchSnapshot()
  })
})
