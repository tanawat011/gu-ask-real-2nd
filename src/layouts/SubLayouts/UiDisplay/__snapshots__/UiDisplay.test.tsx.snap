// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UiDisplayLayout /> match snapshot 1`] = `
<DocumentFragment>
  .c4 {
  z-index: 10;
}

.c2 {
  margin-bottom: 2rem;
}

.c0 {
  position: relative;
  display: grid;
  grid-template-columns: repeat(1,minmax(0,1fr));
  gap: 2rem;
  overflow-x: auto;
  padding-left: 2rem;
  padding-right: 2rem;
  padding-top: 1.5rem;
  padding-bottom: 1.5rem;
}

.c3 {
  grid-column: span 1 / span 1;
  display: none;
}

@media (min-width:1280px) {
  .c0 {
    grid-template-columns: repeat(5,minmax(0,1fr));
  }
}

@media (min-width:1280px) {
  .c1 {
    grid-column: span 4 / span 4;
  }
}

@media (min-width:1280px) {
  .c3 {
    display: block;
  }
}

<div
    class="c0"
    data-testid="ui-display-container-layout"
    id=""
  >
    <div
      class="c1"
    >
      <div>
        <div
          data-testid="title"
        >
          Title
        </div>
      </div>
      <div
        class="c2"
        id=""
      >
        <div
          data-testid="title"
        >
          Title
        </div>
        <table>
          <thead>
            <tr>
              <th>
                1
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                x
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        <div
          data-testid="right-menu-content"
        >
          RightMenuContent
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<UiDisplayLayout /> should can handle scroll event and handle child prop successful 1`] = `
<DocumentFragment>
  .c4 {
  z-index: 10;
  position: fixed;
}

.c2 {
  margin-bottom: 2rem;
}

.c0 {
  position: relative;
  display: grid;
  grid-template-columns: repeat(1,minmax(0,1fr));
  gap: 2rem;
  overflow-x: auto;
  padding-left: 2rem;
  padding-right: 2rem;
  padding-top: 1.5rem;
  padding-bottom: 1.5rem;
}

.c3 {
  grid-column: span 1 / span 1;
  display: none;
}

@media (min-width:1280px) {
  .c0 {
    grid-template-columns: repeat(5,minmax(0,1fr));
  }
}

@media (min-width:1280px) {
  .c1 {
    grid-column: span 4 / span 4;
  }
}

@media (min-width:1280px) {
  .c3 {
    display: block;
  }
}

<div
    class="c0"
    data-testid="ui-display-container-layout"
    id=""
  >
    <div
      class="c1"
    >
      <div>
        <div
          data-testid="title"
        >
          Title
        </div>
      </div>
      <div
        data-testid="fake-component"
      >
        Fake Component
      </div>
      <div
        class="c2"
        id=""
      >
        <div
          data-testid="title"
        >
          Title
        </div>
        <table>
          <thead>
            <tr>
              <th>
                1
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                x
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        <div
          data-testid="right-menu-content"
        >
          RightMenuContent
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
